Address source comments

gassmoeller opened the pull request:
Interesting  experience to create a pull request to one of my own branches :-), but this should make reviewing easier.

This should address all comments from https://github.com/geodynamics/aspect/pull/657. I modified or ignored a few, but tried to clarify why I did things this way, where I did that.

@bangerth: Could you take a look at https://github.com/gassmoeller/aspect/commit/f1e72540de5a8deff83f092860af144d4d6b7dc4. This is my first try to generate tracers before the first timestep. It contains a crude const_cast that I do not like, but the alternative is to write a non-constant find_postprocessor function that is only available for Simulator. Are you ok with the current implementation?

@tjhei: This PR contains two tests that will throw assertions on the tester (tracer_hdf5.prm requires hdf5, tracer_repartition requires deal.II 8.4.pre), should I separate them into another PR and we merge them at a later time? Then this functionality would be untested for the moment however.

gassmoeller commented:
Uh, and ignore the travis warnings below the PR. I once played around a bit, and forgot to disable it again afterwards.

bangerth commented on include/aspect/particle/generator/interface.h:
```diff
@@ -43,6 +43,7 @@ namespace aspect
        */
+#if DEAL_II_VERSION_GTE(8,3,0)
       DeclExceptionMsg (ExcParticlePointNotInDomain,
                         "You requested to generate a particle at a position that "
```
Clever workaround. Alternatively, you could just use `DeclException1` and make the position of the particle something you output.

Thinking about it, this would require knowing the space dimension of the `Point<dim>` argument. That seems unnecessarily complicated. Forget about it.

bangerth replied:
Ah, but then I also forgot that you can prepend everything with `template <int dim>` if you wanted.

But as I said, it's not important. We'll eventually get rid of the `#ifdef` again.

bangerth commented on include/aspect/particle/integrator/interface.h:
```diff
@@ -119,12 +120,16 @@ namespace aspect
+          virtual
+          const void *
+          read_data(const void *data,
+                    const types::particle_index id_num);
```
doc is out of date now wrt `data`.

bangerth replied:
presumably same in all of the other places.

tjhei replied:
and what is the return value?

bangerth commented on include/aspect/particle/output/hdf5.h:
```diff
@@ -24,7 +24,7 @@
 #include <aspect/simulator_access.h>
 
-#include <deal.II/base/data_out_base.h>
+#include <deal.II/numerics/data_out.h>
```
not wrong, but why do you need this change? it pulls in a lot of stuff you don't immediately need.

bangerth commented on include/aspect/particle/world.h:
```diff
@@ -83,6 +83,7 @@ namespace aspect
+         * @param [in] generator The particle generator for this world.
          * @param [in] manager The property manager for this world.
          * @param [in] load_balancing The strategy used to balance the
          * computational load for particle advection across processes.
```
in the function declaration below this comment you pass a lot of `shared_ptr` objects by value. that's almost certainly not wrong, but it doesn't do anything useful unless (i) you call the function with a raw pointer, and (ii) you actually store the resulting shared_ptr object somewhere. In other words, you *capture* a pointer. If that's what you do, it would be easier to read if you just passed the raw pointer. 

But I suspect that that's not what you do, and in that case you'll probably be better off taking these objects as const references.

gassmoeller replied:
I am not exactly sure, what is the right answer to this, I feel like I know when I *can* use unique or shared pointers instead of raw pointers, but not when I *should* use them. 

Let me describe what I am doing: I do not want to add parse_parameters and declare_parameters functions to world.cc as well, so I let the Tracers postprocessor declare and parse input options for the world. Tracers.cc:parse_parameters(prm) reads which plugins to create and creates the plugins, which gives parse_parameters a raw pointer to these objects. These pointers however go out of scope at the end of parse_parameters and after I initialize World with it, therefore I want to transfer ownership of the generated plugins to World during the call to initalize. 

Initially I used raw pointers to hand them over to World, and store them in World. In a review you mentioned I should use unique or shared_ptrs instead (dependings on the purpose), so I went for shared_ptr in Tracers and World (I might have missunderstood you there). Thinking about it again, I probably should have sticked to the raw pointers in tracers.cc:parse_parameters, hand them over as raw pointers to World, and store them as unique pointers in World?

bangerth replied:
Yes, I would say the best strategy is to pass down raw pointers (which already indicates that the *calling* side does not manage them via some kind of smart pointer class) and then the *callee* takes over ownership. You would indicate this in the interface documentation of the function: `@note This function takes over ownership of the objects pointed to by @p generator, @p manager, [...complete list...] objects.`.

Which also means that the list of arguments above is not complete or correctly sorted :-)

tjhei replied:
I don't like the idea of passing ownership using a raw pointer. I think this should be done using unique_ptr.

bangerth replied:
But we do this everywhere, e.g., all of the `create_*_model()` functions return raw pointers.

bangerth commented on include/aspect/particle/world.h:
```diff
@@ -276,6 +283,13 @@ namespace aspect
          * most tracers in the global model.
          */
+        void
+        update_n_global_particles();
```
Comment doesn't match function.

bangerth commented on include/aspect/postprocess/tracers.h:
```diff
@@ -142,31 +151,22 @@ namespace aspect
+         * last_output_time variable.
          */
-        void set_next_data_output_time (const double current_time);
+        void set_last_output_time (const double current_time);
```
nice job aligning with how other postprocessors do it

bangerth commented on source/particle/generator/ascii_file.cc:
```diff
@@ -55,32 +55,17 @@ namespace aspect
-        else
-          Assert(false,ExcNotImplemented());
+        while (in >> particle_position)
+          {
```
nice simplification :-) We can now also run particles in 4d if necessary!

bangerth commented on source/particle/generator/ascii_file.cc:
```diff
@@ -175,7 +160,11 @@ namespace aspect
                                          "a simple text file, with as many columns as spatial "
                                          "dimensions and as many lines as tracers to be generated. "
-                                         "Initial comment lines starting with '#' will be discarded.")
+                                         "Initial comment lines starting with '#' will be discarded."
```
Since this will be translated into latex, you'll have to write this as
```
  Initial comment lines starting with `\\#' will be discarded.
```

bangerth replied:
note also opening backtick.

bangerth commented on source/particle/generator/probability_density_function.cc:
```diff
@@ -186,20 +184,25 @@ namespace aspect
+                iteration++;
               }
-            AssertThrow (num_tries < 100, ExcMessage ("Couldn't generate particle (unusual cell shape?)."));
+            AssertThrow (iteration < 100,
```
This is brittle because you use the same magic number (100) in two different places, the `while` statement and the assertion. Create a variable `const unsigned int max_iterations` or similar that you have to define only once so that the right thing will continue to happen if someone decides to ever change the 100.

I think you have the same code in multiple places -- change it everywhere.

bangerth commented on source/particle/integrator/euler.cc:
```diff
@@ -38,13 +38,18 @@ namespace aspect
-        typename std::multimap<types::LevelInd, Particle<dim> >::iterator it = begin_particle;
-        typename std::vector<Tensor<1,dim> >::const_iterator vel = old_velocities.begin();
+        Assert(std::distance(begin_particle, end_particle) == old_velocities.size(),
+               ExcMessage("The particle integrator expects the velocity vector to be of equal size"
```
missing space at end of line

bangerth commented on source/particle/integrator/rk_2.cc:
```diff
@@ -41,22 +40,32 @@ namespace aspect
-        typename std::vector<Tensor<1,dim> >::const_iterator old_vel = old_velocities.begin();
-        typename std::vector<Tensor<1,dim> >::const_iterator vel = velocities.begin();
+        Assert(std::distance(begin_particle, end_particle) == old_velocities.size(),
+               ExcMessage("The particle integrator expects the old velocity vector to be of equal size"
```
missing space

bangerth commented on source/particle/integrator/rk_2.cc:
```diff
@@ -41,22 +40,32 @@ namespace aspect
+                          "most likely something went wrong in the calling function."));
+
+        Assert(std::distance(begin_particle, end_particle) == velocities.size(),
+               ExcMessage("The particle integrator expects the velocity vector to be of equal size"
```
same here.

by the way, this is inefficient. std::distance on multimap iterators is O(N), so doing the same operation in successive asserts is expensive. in the second assertion, you could just compare the sizes of velocities and old_velocities.

or you could decide not to care because it's just in debug mode :-)

gassmoeller replied:
It is debug mode, I do not care ;-). 
But we do not need to make it slower than necessary, I changed it.

bangerth commented on source/particle/integrator/rk_4.cc:
```diff
@@ -44,33 +40,45 @@ namespace aspect
       {
-        typename std::multimap<types::LevelInd, Particle<dim> >::iterator it = begin_particle;
+        Assert(std::distance(begin_particle, end_particle) == old_velocities.size(),
+               ExcMessage("The particle integrator expects the old velocity vector to be of equal size"
```
missing spaces and other issues as above

bangerth commented on source/particle/interpolator/first_particle.cc:
```diff
@@ -52,7 +52,15 @@ namespace aspect
             const types::LevelInd cell_index = std::make_pair<unsigned int, unsigned int> (cell->level(),cell->index());
+
+            AssertThrow(particles.find(cell_index) != particles.end(),
+                        ExcMessage("At least one cell contained no particles. The 'First"
```
missing space at end of line

bangerth commented on source/particle/output/ascii.cc:
```diff
@@ -42,9 +42,15 @@ namespace aspect
-          std::cout << "ERROR: proc " << Utilities::MPI::this_mpi_process(this->get_mpi_communicator()) << " could not create " << full_filename << std::endl;
+
+        AssertThrow (output,
+                     ExcMessage (std::string("Couldn't open ascii particle output file <"
```
i've change this in other places to say "Could not" to be more formal. Could you make the same change here?

bangerth commented on source/particle/output/ascii.cc:
```diff
@@ -58,36 +64,26 @@ namespace aspect
+                output << property->first << "[" << component_index << "] ";
           }
-        output << "\n";
+        output << '\n';
```
I don't know whether that's actually allowed. Is `\n` really a single-character thing? This may depend on the platform. It's probably safer to leave it in double quotes.

tjhei replied:
Yes, \n is a single character (```0xa```) that you can write as ```'\n'```. It will be translated to whatever byte-sequence the system uses (```\r\n``` on windows).

bangerth commented on source/particle/output/ascii.cc:
```diff
@@ -58,36 +64,26 @@ namespace aspect
+              output << ' ' << properties[i];
 
-            output << "\n";
+            output << '\n';
```
same here

bangerth commented on source/particle/output/interface.cc:
```diff
@@ -100,19 +100,12 @@ namespace aspect
-
-#ifndef DEAL_II_HAVE_HDF5
-            AssertThrow (name != "hdf5",
-                         ExcMessage ("deal.ii was not compiled with HDF5 support, "
```
Any particular reason for taking out this error?

gassmoeller replied:
I moved it into the Constructor of hdf5.cc to keep the interface.cc clean of specific plugin names. It will still throw in the same function and with the same error if somebody tries to use hdf5 without having deal.II compiled with it.

bangerth replied:
Ah, much better indeed.

bangerth commented on source/particle/world.cc:
```diff
@@ -318,10 +336,11 @@ namespace aspect
                              const void *data)
     {
       const unsigned int *n_particles_in_cell = static_cast<const unsigned int *> (data);
```
This variable is poorly named because it doesn't denote a number but a *pointer* into an array of numbers. Maybe `particles_in_cell_ptr` or similar?

bangerth commented on source/postprocess/tracers.cc:
```diff
@@ -297,6 +298,22 @@ namespace aspect
+      const unsigned int n_processes = Utilities::MPI::n_mpi_processes(this->get_mpi_communicator());
+
+      AssertThrow((n_processes == 1) || (CFL_number <= 1.0),
+                  ExcMessage("The current tracer algorithm does not work in"
```
missing space

bangerth commented on source/simulator/initial_conditions.cc:
```diff
@@ -173,6 +174,18 @@ namespace aspect
+
+    // If the tracer postprocessor has been selected
+    if (tracer_postprocessor != 0)
+      tracer_postprocessor->generate_and_initialize_particles();
```
This function looks good. It's not overly elegant to use the `const_cast`, but I think it's a defensible practice here.

I would have named the function without the `_if_selected` and simply noted in the documentation that the function does nothing if no particle postprocessor has been selected

bangerth commented:
YES!

Fix my minor quibbles and generate the final pull request so that we can finally merge this monster!

tjhei commented:
> @tjhei: This PR contains two tests that will throw assertions on the tester (tracer_hdf5.prm requires hdf5, tracer_repartition requires deal.II 8.4.pre), should I separate them into another PR and we merge them at a later time? Then this functionality would be untested for the moment however.

I would prefer we don't add failing tests. Not sure what to do here. I can add hdf5 when I bump the deal.II version but I don't want to require 8.4.pre. We could investigate switching over to the deal.II test support instead of rolling our own, which allows filtering by deal.II version. Another option would be to disable the test for now (for example by renaming the .prm to something else).

tjhei commented on source/particle/world.cc:
```diff
@@ -552,42 +590,42 @@ namespace aspect
-        MPI_Isend(&(num_send_data[i]), 1, MPI_INT, neighbors[i], 0, this->get_mpi_communicator(), &(num_requests[2*i+1]));
-      MPI_Waitall(2*num_neighbors,num_requests,MPI_STATUSES_IGNORE);
-      delete[] num_requests;
+      MPI_Request *n_requests = new MPI_Request[2*n_neighbors];
```
I prefer a ```std::vector<MPI_Request>``` because then you don't need to delete things.

gassmoeller commented:
Ok, I addressed all comments, but left the raw pointer issue unhandled for now (I create raw pointers in tracers.cc, hand them over to World and let it take ownership by using unique_ptr in there). If we want to change this system, we should do it in all interfaces in one pull request to not confuse readers.

I will merge this into the main pull request, and then rebase geodynamics/aspect#411 onto a recent master. Should I squash the 160 commits into 2 (source and tests) or leave them around?

tjhei commented:
Is there valuable history in there? If not I would just go ahead and squash.
